pkgs/development/compilers/gcc/common/configure-flags.nix

```
@@ -114,6 +114,14 @@ let
     ++ lib.optional (!(crossMingw && crossStageStatic))
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+    # cross-built gcc needs to be tricked as if it is built with headers
+    # available in sysroot. For that we need to pass some --with-sysroot= value:
+    #   https://github.com/NixOS/nixpkgs/pull/181802#issuecomment-1186822355
+    # Without this flag --with-native-system-header-dir= has no effect
+    # and target libc features are not detected like stack protector support
+    ++ lib.optional (buildPlatform != hostPlatform && hostPlatform == targetPlatform)
```

```suggestion
    ++ lib.optional (buildPlatform != hostPlatform)
```

Thanks for merging my other suggestion; with that in place we can now simplify this.

---

I did it in a slightly differently by applying it under the same condition as `--with-native-system-header-dir=`.